lib/web/files.go

```diff
@@ -105,102 +105,59 @@ func (h *Handler) transferFile(w http.ResponseWriter, r *http.Request, p httprou
-	// the response, which would corrupt the downloaded file.
-	return nil, nil
-}
-
-type fileTransfer struct {
-	// ctx is a web session context for the currently logged in user.
-	ctx           *SessionContext
-	authClient    auth.ClientI
-	proxyHostPort string
-}
-
-func (f *fileTransfer) download(req fileTransferRequest, httpReq *http.Request, w http.ResponseWriter) error {
-	ctx := httpReq.Context()
-	cmd, err := scp.CreateHTTPDownload(scp.HTTPTransferRequest{
-		RemoteLocation: req.remoteLocation,
-		HTTPResponse:   w,
-		User:           f.ctx.GetUser(),
-	})
+	tc, err := ft.createClient(req, r)
 	if err != nil {
-		return trace.Wrap(err)
-	}
-
-	tc, err := f.createClient(req, httpReq)
-	if err != nil {
-		return trace.Wrap(err)
+		return nil, trace.Wrap(err)
 	}
 
 	if req.webauthn != "" {
-		err = f.issueSingleUseCert(req.webauthn, httpReq, tc)
+		err = ft.issueSingleUseCert(req.webauthn, r, tc)
 		if err != nil {
-			return trace.Wrap(err)
+			return nil, trace.Wrap(err)
 		}
 	}
 
+	ctx := r.Context()
 	if req.fileTransferRequestID != "" {
 		// These values should never exist independently of each other so we can set them at the same time
-		ctx = context.WithValue(ctx, scp.FileTransferRequestID, req.fileTransferRequestID)
-		ctx = context.WithValue(ctx, scp.ModeratedSessionID, req.moderatedSessionID)
+		ctx = context.WithValue(ctx, sftp.FileTransferRequestID, req.fileTransferRequestID)
+		ctx = context.WithValue(ctx, sftp.ModeratedSessionID, req.moderatedSessionID)
 	}
 
-	err = tc.ExecuteSCP(ctx, req.serverID, cmd)
+	err = tc.TransferFiles(ctx, req.login, req.serverID+":0", cfg)
 	if err != nil {
-		return trace.Wrap(err)
+		return nil, trace.Wrap(err)
 	}
 
-	return nil
+	// We must return nil so that we don't write anything to
+	// the response, which would corrupt the downloaded file.
+	return nil, nil
 }
 
-func (f *fileTransfer) upload(req fileTransferRequest, httpReq *http.Request) error {
-	ctx := httpReq.Context()
-	cmd, err := scp.CreateHTTPUpload(scp.HTTPTransferRequest{
-		RemoteLocation: req.remoteLocation,
-		FileName:       req.filename,
-		HTTPRequest:    httpReq,
-		User:           f.ctx.GetUser(),
-	})
-	if err != nil {
-		return trace.Wrap(err)
-	}
-
-	tc, err := f.createClient(req, httpReq)
-	if err != nil {
-		return trace.Wrap(err)
-	}
-
-	if req.webauthn != "" {
-		err = f.issueSingleUseCert(req.webauthn, httpReq, tc)
-		if err != nil {
-			return trace.Wrap(err)
-		}
-	}
-
-	if req.fileTransferRequestID != "" {
-		// These values should never exist independently of each other so we can set them at the same time
-		ctx = context.WithValue(ctx, scp.FileTransferRequestID, req.fileTransferRequestID)
-		ctx = context.WithValue(ctx, scp.ModeratedSessionID, req.moderatedSessionID)
-	}
-
-	err = tc.ExecuteSCP(ctx, req.serverID, cmd)
-	if err != nil {
-		return trace.Wrap(err)
-	}
-
-	return nil
+type fileTransfer struct {
+	// ctx is a web session context for the currently logged in user.
+	ctx           *SessionContext
```

Comment:
Let's name this sctx. We should reserve ctx for context.Context values

Reply:
Done